### PR TITLE
test: refactor expect

### PR DESCRIPTION
**Pull request by tomberek**
## Proposed Changes

- bring bats tests back to all systems
- Remove deprecated tests
- improve expect test robustness
- start implementing pure tests (example provided)

## Release Notes
N/A

### CODE REVIEW

File: cli/tests/environment-install.bats [hunk 1]
```diff
@@ -145,8 +145,7 @@ teardown() {
   mkdir 2
   "$FLOX_BIN" init --dir 2
 
-  SHELL=bash run expect -d "$TESTS_DIR/install/prompt-which-environment.exp"
-  assert_success
+  SHELL=bash run -0 expect -d "$TESTS_DIR/install/prompt-which-environment.exp"
```
Comment (ysndr) [inline cli/tests/environment-install.bats:148]: no assert_success because expect ends with `exit 0` anyway?

File: cli/tests/activate/hook.exp [hunk 2]
```diff
@@ -3,23 +3,13 @@
 
 set dir [lindex $argv 0]
 set flox $env(FLOX_BIN)
-
-set timeout 5
-set cmd "$flox activate --dir $dir"
```
Comment (ysndr) [inline cli/tests/activate/hook.exp:8]: we don't need to `set cmd`?

File: cli/tests/activate/hello.exp [hunk 3]
```diff
@@ -6,62 +6,23 @@
...
-    timeout {
-      puts stderr "Reached timeout running '$cmd'"
-      exit 1
-    }
-}
-
-# activation is the only thing that should take very long so set a low timeout
-set timeout 1
+expect "Building"
+expect -re "flox .*\\\[project-\\d+\\\]" {}
```
Comment (ysndr) [inline cli/tests/activate/hello.exp:16]: I'm a bit confused as to when we are using expect to "expect" certain strings and when we `assert_output` the _output_ in bats.
this one seems to do the former while `cli/tests/activate/hook.exp` is doing the latter

**Review by ysndr — COMMENTED** [review]
I'm not sure if I'm qualified to approve the expect stuff but i left some questions around

File: pkgs/flox-cli/default.nix [hunk 4]
```diff
@@ -34,17 +34,19 @@
...
+      ! builtins.elem path (map (
+          f: "${./../../cli}/${f}"
+        ) [
+          "flake.nix"
+          "flake.lock"
+          "pkgs"
+          "checks"
+          "tests"
+          "shells"
+          "target"
```
[the remark this answers is not in the document]
Comment (tomberek) [reply pkgs/flox-cli/default.nix:48]: Possibly, yes. Or use the more advanced filset library recently added to Nixpkgs.

File: pkgs/flox-cli/default.nix [hunk 5]
```diff
@@ -34,17 +34,19 @@
 }: let
   flox-src = builtins.path {
     name = "flox-src";
-    path = ./../../cli;
+    path = "${./../../cli}";
```
[the remark this answers is not in the document]
Comment (tomberek) [reply pkgs/flox-cli/default.nix:37]: Yes. Or at least we have to ensure the function is working as expected - the way it was there function was comparing a string to a path as well as having mismatched named between "source" and "cli".

File: cli/tests/environment-install.bats (hunk 1 above)
Comment (tomberek) [reply cli/tests/environment-install.bats:148]: The "-0" handles that check.

File: cli/tests/activate/hello.exp (hunk 3 above)
Comment (tomberek) [reply cli/tests/activate/hello.exp:16]: Before sending commands we need to ensure the spawn is up and running. Possible the second expect here is not needed and can be removed.

Ideally we have minimal checking, especially for the finicky prompt stuff, in expect.

File: cli/tests/activate/hook.exp (hunk 2 above)
Comment (tomberek) [reply cli/tests/activate/hook.exp:8]: cmd was used to enhance the error output, but made the flow harder the understand and easy to mess up.

**Comment by tomberek** [comment]
currently rebased on top of: https://github.com/flox/flox/pull/607

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -330,14 +330,26 @@ jobs:
...
+          else
+            nix build -L \
               --accept-flake-config \
               --no-update-lock-file \
               --print-out-paths \
                 '.#flox-pkgdb' \
                 '.#flox-env-builder' \
-                '.#flox-cli' \
-                '.#flox'
+                '.#flox-cli'
```
Comment (garbas) [inline .github/workflows/ci.yml:351]: We "sadly" still need to built also .#flox`but it is going to be a wrapper soon.

```suggestion
                '.#flox-cli' \
                '.#flox'
```

File: pkgs/flox-cli-tests/default.nix [hunk 7]
```diff
@@ -31,7 +31,7 @@
   which,
   writeShellScriptBin,
   PROJECT_NAME ? "flox-cli-tests",
-  PROJECT_TESTS_DIR ? ./../../cli/tests,
+  PROJECT_TESTS_DIR ? "${./../..}/cli/tests",
```
Comment (garbas) [inline pkgs/flox-cli-tests/default.nix:34]:
```suggestion
  PROJECT_TESTS_DIR ? "./../../cli/tests",
```

**Review by garbas — APPROVED** [review]

**Pull request merged**